### ISSUE TITLE
Queries with joins across multiple tables with the same column name cause values to be overwritten when outputted as an object

### ISSUE THREAD

**Issue by lostless13** [oracle/node-oracledb#1250]
Example Scenario:
Table_1 -> Columns: ID, NAME, QUANTITY
Table_2 -> Columns: ID_FOREIGN_KEY, PROPERTY, QUANTITY

Query:
```
SELECT * FROM Table_1 LEFT JOIN Table_2 on Table_1.ID == Table_2.ID_FOREIGN_KEY
```

Expected rows array in return object:
```
[{"ID": 1,
"NAME": "Test",
 "QUANTITY":1,  
"ID_FOREIGN_KEY": 1, 
"PROPERTY": "Test prop",
"QUANTITY_1": 3},  <-- appended 1 to avoid collision with existing quantity property
 ....]
```
Actual rows array in return object:
```
[{"ID": 1, 
"NAME": "Test", 
"QUANTITY":3,  
"ID_FOREIGN_KEY": 1, 
"PROPERTY": "Test prop"}, 
....]
```

Note that the quantity column in the returned row object was overwritten as opposed to creating an new object property and appending a value to the end as is customary in many interfaces.

**Comment by cjbj** [comment]
For future readers, the work-arounds are (i) use a column alias in the query - which is easy but only when you have control over the SQL (ii) Use OUT_FORMAT_ARRAY.

This is a long-known feature in Oracle-land with multi-table queries.

Can you give examples of the 'many interfaces' that append values?

**Comment by lostless13** [comment]
Thank you. Interfaces may have been a poor choice of words but I was referring to GUIs. I've been using DB Forge Studio for this project which has the appending behavior. I haven't used SQL Developer in sometime but I believe it also will append values in the case of duplicate column names.

Here's a helper function I wrote to handle this situation for me and might be helpful to others. It takes the results of a query in the OUT_FORMAT_ARRAY and converts each row into an object with unique column names. It modifies the result in place which may not be palatable to some and is only lightly tested so proceed at your own risk. I also included the option to remove null values as this was of value in my applications.
```
function formatOracleArrayAsObject(result, removeNulls = false) {
    if (!result.hasOwnProperty("metaData") || !result.hasOwnProperty("rows")) {
        return
    }

    let metaData = result.metaData
    let rows = result.rows

    // update meta data array by appending a counter value to repeated column names
    let colNameDict = {}
    for (let col of metaData) {
        if (colNameDict.hasOwnProperty(col.name)) {
            col.name = `${col.name}_${colNameDict[col.name]++}`
        } else {
            colNameDict[col.name] = 1
        }
    }

    // convert row arrays to objects with named properties using updated column names
    for (let i = 0; i < rows.length; i++) {
        rows[i] = rows[i].reduce((acc, value, j)=>{
            if (value || !removeNulls) {
                acc[metaData[j].name] = value
            }
            
            return acc
        }, {})
    }
}
```

**Comment by cjbj** [comment]
Thanks for sharing.

The general idea does seem to have a lot of overhead compared with using a column alias e.g. `select dummy as mycol from dual`.  Hopefully future readers will make an informed choice about their requirements!

**Comment by cjbj** [comment]
We'll look at this for node-oracledb 5.1

**Comment by cjbj** [comment]
Code for this was merged to the master branch. See https://oracle.github.io/node-oracledb/INSTALL.html#-37-node-oracledb-installation-from-source-code for how to build it.

**Comment by cjbj** [comment]
This was released in node-oracledb 5.1.
